### ISSUE TITLE
Compatibility with react-redux@^6.0.0

### ISSUE THREAD

**Issue by dalwlad** [reduxjs/redux-devtools#417]
Because `redux-devtools@^6.0.0` is no longer supporting passing the `store` as a prop to connected components, using `redux-devtools` is no longer working (when [manually integrating as detailed here](https://github.com/reduxjs/redux-devtools/blob/master/docs/Walkthrough.md#manual-integration)), but instead throwing an error

```
Uncaught Error: Passing redux store in props has been removed and does not do anything. To use a custom Redux store for specific components,  create a custom React context with React.createContext(), and pass the context object to React-Redux's Provider and specific components like:  <Provider context={MyContext}><ConnectedComponent context={MyContext} /></Provider>. You may also pass a {context : MyContext} option to connect
    at invariant (browser.js:38)
    at new Connect (connectAdvanced.js:137)
    at constructClassInstance (react-dom.development.js:12484)
    at updateClassComponent (react-dom.development.js:14255)
    at beginWork (react-dom.development.js:15082)
    at performUnitOfWork (react-dom.development.js:17820)
    at workLoop (react-dom.development.js:17860)
    at HTMLUnknownElement.callCallback (react-dom.development.js:149)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:199)
    at invokeGuardedCallback (react-dom.development.js:256)
```

**Comment by zalmoxisus** [comment]
Thanks for the report! I guess it's complaining on [this line](https://github.com/reduxjs/redux-devtools/blob/master/src/createDevTools.js#L60), but don't see it in your stack trace.

I cannot get the warning with [counter](https://github.com/reduxjs/redux-devtools/tree/master/examples/counter) example, but I see that the store is not found in context with `redux-devtools@6`. Adding it to `DevTools` props solves the issue:
```diff
import React, { Component } from 'react';
import { Provider } from 'react-redux';
import CounterApp from './CounterApp';
import DevTools from './DevTools';

export default class Root extends Component {
  render() {
    const { store } = this.props;
    return (
      <Provider store={store}>
        <div>
          <CounterApp />
-          <DevTools />
+          <DevTools store={store} />
        </div>
      </Provider>
    );
  }
}
```

**Comment by dalwlad** [comment]
@zalmoxisus `DevTools` doesn't seem to work. I get the runtime exception and nothing renders in the popup. The exception is thrown in the parent window and the popup is just a blank page.

**Comment by evang522** [comment]
Just updated to react-redux 6.0.0 and my team are also experiencing this issue. Typically our development App root looks like this: 

```
      <Provider store={store}>
             // ... some more of our application
         <DevTools/>
        </Provider>

```
Here we receive the console error `Redux DevTools could not render. You must pass the Redux store to <DevTools> either as a "store" prop or by wrapping it in a <Provider store={store}>.` 
(And dev tools do not work at all) 

But when passing the store into the DevTools component, like so: 
```
        <Provider store={store}>
             // ... some more of our application
         <DevTools store={store}/>
        </Provider>
```

We receive the error: 

`Uncaught Error: Passing redux store in props has been removed and does not do anything. To use a custom Redux store for specific components,  create a custom React context with React.createContext(), and pass the context object to React-Redux's Provider and specific components like:  <Provider context={MyContext}><ConnectedComponent context={MyContext} /></Provider>. You may also pass a {context : MyContext} option to connect`


Relevant package JSON versions: 

```
    "react-redux": "^6.0.0",
    "react": "^16.6.3",
    "redux": "^4.0.1",
    "redux-devtools": "^3.4.2",
    "redux-devtools-dock-monitor": "^1.1.3",
    "redux-devtools-log-monitor": "^1.4.0",
```

Dev Tools Implementation: 

```
export default createDevTools(
    <DockMonitor defaultIsVisible={false} visibleOnLoad={false} toggleVisibilityKey="ctrl-h" changePositionKey="ctrl-w">
        <LogMonitor/>
    </DockMonitor>
);
```

Store Enhancer Implementation: 

```
export function configureStore(initialState) {
    return createStore(
        rootReducer,
        initialState,
        compose(
            middleware,
            DevTools.instrument(),
        )
    );
}

```


Anything we're missing you can point out or is this an issue with compatibility?

**Comment by GuillaumeCisco** [comment]
I have the exact same error @evang522 
I created a codesandbox example with the devTools displaying but not working: https://codesandbox.io/s/14zw28lz2q

Issue is the same as some discussed in https://github.com/reduxjs/react-redux/issues/1126 and https://github.com/reduxjs/react-redux/issues/1133
I'm close to a solution but need help.

I think this issue will interest @Ephem and @markerikson

**Comment by zalmoxisus** [comment]
@GuillaumeCisco should that example display something else? I see DevTools is working there and showing `state: undefined`, which is what it should be.

**Comment by Ephem** [comment]
I'm certain this issue is unrelated to both react-redux#1126 and react-redux#1133.

Just to be clear to any new readers of this issue, `redux-devtools` is currently not expected to be compatible with v6 of `react-redux`. If you check the `peerDependencies` in `package.json`:

```json
  "peerDependencies": {
    "react": "^0.14.9 || ^15.3.0 || ^16.0.0",
    "react-redux": "^4.0.0 || ^5.0.0",
    "redux": "^3.5.2 || ^4.0.0"
  },
```

Any problems that you get are to be expected, so this issue should be about finding out what is needed to be compatible with `react-redux` v6.

That said, v6 uses the new "official" React-context, while v5 and back uses the old one, so I think the reason the store is not found on the context can be tracked [to this line](https://github.com/reduxjs/redux-devtools/blob/master/packages/redux-devtools/src/createDevTools.js#L13).

I don't really have much knowledge about the `redux-devtools`, but I can imagine there are a lot of considerations to account for here, to start:

* Is it possible to be compatible with both v5 and v6 or is this a breaking?
* The new `ReactReduxContext` is not guaranteed to be used, since it is possible to pass in your own context, devTools must account for this.
* v6 also places `storeState` on the context. This corresponds to the store state at the start of the rendering pass, so it might diff with `store.getState()` because that always gets the latest state. How will this affect the devTools?

Since v6 is pretty different from v5 behind the scenes, I imagine I've missed things, someone better informed can probably fill in the blanks? :)

**Comment by markerikson** [comment]
No, anything that tries to access the store directly can't be compatible with both v5 and v6 at the same time, [per the v6 release notes](https://github.com/reduxjs/react-redux/releases/tag/v6.0.0).  

It looks like this package renders:

`<ConnectedMonitor {...monitorProps} store={this.liftedStore} />`

That would probably have to change to wrap it in `<Provider store={liftedStore}><ConnectedMonitor {...monitorProps} /></Provider>`.  That said, it also would override the store for anything else inside it, but I don't think that `<DevTools>` accepts any children anyway.

**Comment by zalmoxisus** [comment]
It will be breaking anyway, as passing the `store` to DevTools in props, as an alternative for context, was expected as per #176. I see @GuillaumeCisco has already a working solution in https://codesandbox.io/s/14zw28lz2q and it looks to work as expected (state is `undefined` as there are no reducers to set/change it).

**Comment by Ephem** [comment]
After skimming the source I do believe that it is possible to keep supporting store as a prop to DevTools. Instead of DevTools passing that prop on to the ConnectedMonitor, it can create a valid ”new” redux-context and wrap the monitor in that as per Marks suggestion above.

In other words, DevTools reads store either from props or context, but always passes it on via context.

(Also, if I’m reading those release notes correctly it should be possible to just pass that context to the ConnectedComponent directly instead of wrapping in new Provider? Since we can still get a single context by defining contextType on the class, I think we can get away without using either a Context.Consumer or a Context.Provider here?)

**Comment by Ephem** [comment]
Without knowing at all what I’m talking about here, but maybe in theory it would be possible for the devtools to detect if ReactReduxContext is present on the export from react-redux and behave differently for the different versions to keep supporting the old versions in same package?

**Comment by zalmoxisus** [comment]
Good points @Ephem! I'll publish tomorrow a patch which will work only when store is passed to DevTools via props, passing then wrapping connected monitor into a `Provider` like @markerikson suggested. That shouldn't be breaking. If someone wants to work on a PR, getting it also from context that would be welcome.

I figured why my example in https://github.com/reduxjs/redux-devtools/issues/417#issuecomment-445830339 was working, I had `react-redux@6` in main project, but `react-redux@5` for `redux-devtools` dependency. So if someone needs a quick dirty (you shouldn't include DevTools component in production anyway) just do: `cd ./node_modules/redux-devtools && npm i react-redux@5`.

**Comment by GuillaumeCisco** [comment]
As you can see in my codesandbox: https://codesandbox.io/s/14zw28lz2q
Errors are no more displayed, the Redux DevTools Dock Monitor show itself, but it does not react to any actions (ctr+h, ctrl+m, ctrl+q, resizing width from the mouse). It is related to https://github.com/reduxjs/react-redux/issues/1133 in a way that the action is called, the reducer is called, but `mapStateToProps` is not called and the render is not triggered.

As you can see in the sandbox, I define the devTools as:
```javascript
import React from "react";

// Exported from redux-devtools
import { createDevTools } from "../redux-devtools";

// Monitors are separate packages, and you can make a custom one
import LogMonitor from "redux-devtools-log-monitor";
import DockMonitor from "redux-devtools-dock-monitor";
import SliderMonitor from "redux-slider-monitor";

// createDevTools takes a monitor and produces a DevTools component
const DevTools = createDevTools(
  // Monitors are individually adjustable with props.
  // Consult their repositories to learn about those props.
  // Here, we put LogMonitor inside a DockMonitor.
  // Note: DockMonitor is visible by default.
  <DockMonitor
    toggleVisibilityKey="ctrl-h"
    changePositionKey="ctrl-q"
    changeMonitorKey="ctrl-m"
    defaultIsVisible
  >
    <LogMonitor theme="tomorrow" />
    <SliderMonitor keyboardEnabled />
  </DockMonitor>
);

export default DevTools;

```
Pay attention at how I refer to `DevTools`:
```javascript
import { createDevTools } from "../redux-devtools";
```

I rewrote the package for using the `withRedux` component we talked about in reduxjs/react-redux#1126.

`createDevTools.js`
```javascript
import React, { Children, Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import instrument from "redux-devtools-instrument";
import withRedux from "./withRedux";

export default function createDevTools(children) {
  console.log("createDevTools");

  const monitorElement = Children.only(children);
  const monitorProps = monitorElement.props;
  const Monitor = monitorElement.type;
  const mapStateToProps = state => {
    console.log("mapStateToProps devtools");
    return state;
  };
  const ConnectedMonitor = withRedux(connect(mapStateToProps)(Monitor));

  class DevTools extends Component {
    render() {
      console.log("render DevTools ");

      return <ConnectedMonitor {...monitorProps} />;
    }
  }

  DevTools.contextTypes = {
    store: PropTypes.object
  };

  DevTools.instrument = options =>
    instrument(
      (state, action) => Monitor.update(monitorProps, state, action),
      options
    );

  return DevTools;
}
```
`withRedux.js`
```javascript
import React, { Component } from "react";
import { ReactReduxContext } from "react-redux";

export default function withRedux(WrappedComponent) {
  class WithRedux extends Component {
    render() {
      console.log("withRedux devtools");
      return (
        <ReactReduxContext.Consumer>
          {({ store, ...reduxContext }) => {
            console.log(store.liftedStore.getState());

            return (
              <ReactReduxContext.Provider
                value={{
                  store: store.liftedStore,
                  storeState: store.liftedStore.getState()
                }}
              >
                <WrappedComponent {...this.props} />
              </ReactReduxContext.Provider>
            );
          }}
        </ReactReduxContext.Consumer>
      );
    }
  }

  return WithRedux;
}
```
Here I pass the liftedStore as the store, as we did before in the constructor of `createDevTools` https://github.com/reduxjs/redux-devtools/blob/master/packages/redux-devtools/src/createDevTools.js#L59
And I manually update the state for reflecting the last one from `store.liftedStore.getState()`

I've put some console.log so you can see what's going on from the console to help debugging.

I hope all these informations will help making this project work with react-redux 6.0.0 without breaking compatibility with 5.x ;)

**Comment by GuillaumeCisco** [comment]
In an effort to provide more debugging tools, I created a sandbox with versions of `redux-devtools`, `redux-devtools-dock-monitor`, `redux-devtools-log-monitor` and `redux-devtools-slider-monitor` modified to include some `console.log/console.debug` lines.

It is easier to see what's is going on, actions and reducers are called, but the `mapStateToProps` is not.
Adding what @markerikson suggested in reduxjs/react-redux#1126 a `setTimeOut` in `componentDidMount`, we can see the state is well updated.

Codesandbox is available here: https://codesandbox.io/s/j2ym2pnopw
Do not hesitate to uncomment code in `withRedux.js` file for seeing component correctly taking account of the updated state.

Hope it will help ;)

**Comment by zalmoxisus** [comment]
#427 should fix it, wrapping in `Provider` only for `react-redux@6` and if `store` is not passed in props, will also wrap into `ReactReduxContext.Consumer` to get it from context. So it's not breaking for lower versions.

It's published as `redux-devtools@3.5.0-beta` if someone wants to give it a try.

**Comment by zalmoxisus** [comment]
Published as `redux-devtools@3.5.0`. Thanks to everybody for helping with it!
